drv/cosmo-hf/src/hf.rs

```diff
@@ -0,0 +1,209 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use drv_hash_api::SHA256_SZ;
+use drv_hf_api::{
+    HfDevSelect, HfError, HfMuxState, HfPersistentData, HfProtectMode,
+};
+use idol_runtime::{
+    LeaseBufReader, LeaseBufWriter, Leased, LenLimit, NotificationHandler,
+    RequestError, R, W,
+};
+use ringbuf::ringbuf_entry_root as ringbuf_entry;
+use userlib::{task_slot, RecvMessage};
+
+use crate::{ServerImpl, Trace, PAGE_SIZE_BYTES, SECTOR_SIZE_BYTES};
+
+task_slot!(HASH, hash_driver);
+
+impl idl::InOrderHostFlashImpl for ServerImpl {
+    fn read_id(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<[u8; 20], RequestError<HfError>> {
+        Ok(self.flash_read_id())
+    }
+
+    fn capacity(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<usize, RequestError<HfError>> {
+        todo!()
+    }
+
+    /// Reads the STATUS_1 register from the SPI flash
+    fn read_status(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError<HfError>> {
+        Ok(self.read_flash_status())
+    }
+
+    fn bulk_erase(
+        &mut self,
+        _: &RecvMessage,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if !matches!(protect, HfProtectMode::AllowModificationsToSector0) {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        todo!()
+    }
+
+    fn page_program(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+        data: LenLimit<Leased<R, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr as usize / SECTOR_SIZE_BYTES == 0
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_page_program(
+            addr,
+            LeaseBufReader::<_, 32>::from(data.into_inner()),
+        )
+        .map_err(|()| RequestError::went_away())
+    }
+
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        offset: u32,
+        dest: LenLimit<Leased<W, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        self.flash_read(
+            offset,
+            LeaseBufWriter::<_, 32>::from(dest.into_inner()),
+        )
+        .map_err(|_| RequestError::went_away())
+    }
+
+    fn sector_erase(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr as usize / SECTOR_SIZE_BYTES == 0
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_sector_erase(addr);
+        Ok(())
+    }
+
+    fn get_mux(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfMuxState, RequestError<HfError>> {
+        todo!()
```

Comment:
There's a lot of TODOs here providing remote panic opportunities. Are these operations going to make sense for Cosmo? I ask because, if not, we should have two different idl files, since we can't cfg things in IDL files at the moment.

Reply:
From talking with Nathanael, we do think that the existing API makes sense for Cosmo!

There are PRs coming down the line that implement these functions, e.g. #1853, but I didn't want to do too much in the MVP PR.